### PR TITLE
Create ORT opschema library

### PR DESCRIPTION
**Pull request by gramalingam**
**Description**: Create a library of ORT custom ops, for use in other applications

**Motivation and Context**
The ONNX-to-MLIR translator needs access to the schemas of custom ops to translate ONNX models containing these ops into MLIR. A standalone translator can link the library of the opschemas for this purpose.

Design choice: Currently, we just expose a single register method that registers all the custom ops (including both inference and training). If desired, we can extend this in the future to register specific subsets.

### CODE REVIEW

File: cmake/onnxruntime_opschema_lib.cmake [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+
+set (CONTRIB_OPS_DIR ${ONNXRUNTIME_ROOT}/core/graph/contrib_ops)
+set (TRAINING_OPS_DIR ${ORTTRAINING_ROOT}/orttraining/core/graph)
+
+file(GLOB_RECURSE contrib_ops_schema_src
+   "${CONTRIB_OPS_DIR}/*.cc"
+   "${TRAINING_OPS_DIR}/training_op_defs.cc"
+)
+
+add_library(ort_opschema_lib ${contrib_ops_schema_src})
```
Comment (snnn) [inline cmake/onnxruntime_opschema_lib.cmake:11]: A static library? How will it be used?

File: cmake/onnxruntime_opschema_lib.cmake [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+
+# ${CMAKE_CURRENT_BINARY_DIR} is so that #include "onnxruntime_config.h" is found
+target_include_directories(ort_opschema_lib PRIVATE ${ONNXRUNTIME_ROOT} ${ORTTRAINING_ROOT} ${CMAKE_CURRENT_BINARY_DIR})
+onnxruntime_add_include_to_target(ort_opschema_lib onnxruntime_common onnx onnx_proto protobuf::libprotobuf flatbuffers)
```
Comment (snnn) [inline cmake/onnxruntime_opschema_lib.cmake:15]: Is it possible to make it not depend on onnxruntime_common?

File: cmake/onnxruntime_opschema_lib.cmake (hunk 1 above)
Comment (snnn) [reply cmake/onnxruntime_opschema_lib.cmake:11]: Usually it's hard to distribute a static library binary. For example, GCC has two  So would you please tell us more about your plan?

File: cmake/onnxruntime_opschema_lib.cmake [hunk 3]
```diff
@@ -0,0 +1,28 @@
+# Schema library (for contrib ops and training ops)
```
Comment (pranavsharma) [inline cmake/onnxruntime_opschema_lib.cmake:1]: Needs copyright headers.

File: samples/c_cxx/opschema_lib_use/main.cc [hunk 4]
```diff
@@ -0,0 +1,33 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
```
Comment (pranavsharma) [inline samples/c_cxx/opschema_lib_use/main.cc:1]: Samples don't get run as part of CI. If you want this functionality tested, it should be part of our unit tests.

File: cmake/onnxruntime_opschema_lib.cmake (hunk 1 above)
Comment (snnn) [reply cmake/onnxruntime_opschema_lib.cmake:11]: Talked with @gramalingam offline. We will try this approach first. The users of this library need to make sure they will use the same onnx version/protobuf version as onnx runtime.

File: samples/c_cxx/opschema_lib_use/main.cc (hunk 4 above)
Comment (gramalingam) [reply samples/c_cxx/opschema_lib_use/main.cc:1]: Thanks. Yes, I was wondering if this should be a part of the CI or not. At the moment, its main purpose is to act as a sample for users who want to link in the library, which is why I decided to leave it as a sample. If needed, we can turn it into a CI unit test.

File: cmake/onnxruntime_opschema_lib.cmake (hunk 2 above)
Comment (gramalingam) [reply cmake/onnxruntime_opschema_lib.cmake:21]: Unfortunately, there were some dependencies. One is GetStackTrace(), which is used in common macros. Any suggestions?

File: cmake/onnxruntime_opschema_lib.cmake (hunk 3 above)
Comment (gramalingam) [reply cmake/onnxruntime_opschema_lib.cmake:4]: Done.

File: cmake/onnxruntime_opschema_lib.cmake [hunk 5]
```diff
@@ -13,7 +13,7 @@ file(GLOB_RECURSE contrib_ops_schema_src
 
 onnxruntime_add_static_library(ort_opschema_lib ${contrib_ops_schema_src})
 
-set (OPSCHEMA_LIB_DEPENDENCIES onnxruntime_mlas onnxruntime_common onnxruntime_util onnx onnx_proto protobuf::libprotobuf flatbuffers)
+set (OPSCHEMA_LIB_DEPENDENCIES onnxruntime_util onnxruntime_common onnxruntime_mlas onnx onnx_proto protobuf::libprotobuf flatbuffers)
```
Comment (snnn) [inline cmake/onnxruntime_opschema_lib.cmake:16]: onnxruntime_mlas should be on the left side of onnxruntime_common. 

Left can depends on right, not vice versa.

File: cmake/onnxruntime_opschema_lib.cmake (hunk 2 above)
Comment (snnn) [reply cmake/onnxruntime_opschema_lib.cmake:26]: Sorry I don't have any now. We may reconsider it later.

**Review by snnn — APPROVED** [review]
LGTM.

File: cmake/CMakeLists.txt [hunk 6]
```diff
@@ -159,6 +159,9 @@ option(onnxruntime_ENABLE_BITCODE "Enable bitcode for iOS only" OFF)
 # build eager mode
 option(onnxruntime_ENABLE_EAGER_MODE "build ort eager mode")
 
+# build separate library of schemas of (custom) ops used by ORT (for ONNX to MLIR translation)
+option(onnxruntime_BUILD_OPSCHEMA_LIB "Build op schema library" ON)
```
Comment (snnn) [inline cmake/CMakeLists.txt:163]: I feel it is ok to remove this option and turn it always ON.  Your change has no impact to the core runtime.

File: samples/c_cxx/opschema_lib_use/dummy.cc [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+
+namespace math {
+
+uint16_t floatToHalf(float f) { return 0; }
+
+}  // namespace math
+
+}  // namespace onnxruntime
+
+size_t MlasNchwcGetBlockSize() { return 0; }
```
Comment (snnn) [inline samples/c_cxx/opschema_lib_use/dummy.cc:21]: Interesting, the schemas depends on "MlasNchwcGetBlockSize"?  Later on probably we can refactor the code and make it better. In my opinion, "abort()" is a bit better than "return 0" in such cases.

File: cmake/onnxruntime_opschema_lib.cmake (hunk 5 above)
Comment (gramalingam) [reply cmake/onnxruntime_opschema_lib.cmake]: I have removed all these dependencies. I changed the ordering above based on some existing cmake files, so the existing cmake files may not all be using the same order and may require some fixes. E.g., see: https://github.com/microsoft/onnxruntime/blob/2f2aaf2cf6151e75c4023fc2d0c9d044f5cd212c/cmake/onnxruntime_unittests.cmake#L469

**Review by snnn — DISMISSED** [review]

**Review by souptc — APPROVED** [review]

**Pull request merged**